habitat_code/ramachandran_code_for_review.py

```diff
@@ -0,0 +1,511 @@
+# 
+# <img src="https://upload.wikimedia.org/wikipedia/commons/7/7e/BoulderBearPeak.jpg" alt="Image of Boulder, Colorado" width="400">
+# 
+# > Sources:
+# >
+# > [1] https://bouldercolorado.gov/government/departments/forestry/about#main-content
+# >
+# > [2] https://storymaps.arcgis.com/stories/0cb784ee805144428f914f904a0bb367
+
+# %% [markdown]
+# ## STEP 1: Set up your analysis
+
+# %%
+import cartopy.crs as ccrs
+from census import Census
+import earthpy as et
+import geoviews as gv
+import geopandas as gpd
+import getpass
+
+import hvplot
+import holoviews as hv
+import hvplot.pandas
+import hvplot.xarray
+
+import io
+
+import matplotlib.pyplot as plt
+import numpy as np
+import os
+
+import pystac_client
+import pandas as pd
+
+import requests
+import rioxarray as rxr
+from rioxarray.merge import merge_arrays
+
+
+import shapely
+
+import sklearn
+from sklearn.model_selection import train_test_split
+from sklearn.linear_model import LinearRegression
+from sklearn.metrics import mean_squared_error, r2_score
+
+import us
+
+import xarray as xr
+
+# %%
+data_dir = os.path.join(et.io.HOME, et.io.DATA_NAME, 'boulder-greenspace')
+
+if not os.path.exists(data_dir):
+    os.makedirs(data_dir)
+
+
+# %%
+%%bash
+ls ~/earth-analytics/data/boulder-greenspace
+
+# %%
+# Download City of Boulder boundary
+
+boundary_url = ("https://gis.bouldercolorado.gov/ags_svr1/rest/services/plan/CityLimits/MapServer/0/query?outFields=*&where=1%3D1&f=geojson")
+
+boundary_path = os.path.join(data_dir, 'city_boundary.geojson')
+
+if os.path.exists(boundary_path):
+
+    boundary_gdf = gpd.read_file(boundary_path)
+    print("Data is already downloaded.")
+
+else:
+    # Mimic web browser
+    user_agent = (
+        'Mozilla/5.0 (X11; Linux x86_64; rv:60.0) '
+        'Gecko/20100101 Firefox/81.0'
+    )
+
+    # Download GEOJSON
+    r = requests.get(url=boundary_url, headers={'User-Agent': user_agent})
+
+    # Read GeoJSON data into a GeoDataFrame
+    boundary_geojson_data = r.json()
+
+    # Cache
+    boundary_gdf = (gpd.GeoDataFrame
+                    .from_features(boundary_geojson_data['features'])
+                    .to_file(boundary_path, driver="GeoJSON")
+                    )
+    
+    # Read
+    boundary_gdf = gpd.read_file(boundary_path)
+    print("Data downloaded and loaded.")
+
+
+
+# %%
+boundary_gdf
+
+# %%
+boundary_gdf.to_crs("4326").plot()
+
+# %%
+# Download census tracts
+print(us.states.CO.fips)
+
+colorado_tracts_url = ("https://www2.census.gov/geo/tiger/TIGER2023/"
+           "TRACT/tl_2023_08_tract.zip")
+
+colorado_tracts_path = os.path.join(data_dir, 'colorado_census_tracts.geojson')
+
+if os.path.exists(colorado_tracts_path):
+    colorado_tracts = gpd.read_file(colorado_tracts_path).to_crs("4326")
+    print("Data is already downloaded.")
+else:
+    gpd.read_file(colorado_tracts_url).to_file(colorado_tracts_path, driver="GeoJSON")
+    colorado_tracts_gdf = gpd.read_file(colorado_tracts_path).to_crs("4326")
```

Comment:
This isn't super-repetitive, but it caught my eye because it would be really easy to, say, change the CRS in one spot and not another. In fact, it looks like your code might break if you have to re-download because you have a different variable name (colorado_tracts_gdf vs. colorado_tracts). So watch out for this!